OpenRA.Game/Primitives/PlayerDictionary.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	/// <summary>
+	/// Provides a mapping of players to values, as well as fast lookup by player index.
+	/// </summary>
+	public struct PlayerDictionary<T> : IReadOnlyList<T>, IReadOnlyDictionary<Player, T> where T : class
+	{
+		readonly T[] valueByPlayerIndex;
+		readonly Dictionary<Player, T> valueByPlayer;
+
+		public PlayerDictionary(World world, Func<Player, T> valueFactory)
+		{
+			var players = world.Players;
+			if (players.Length == 0)
+				throw new InvalidOperationException("The players in the world have not yet been set.");
+
+			// The world players never change, so we can safely maintain an array of values.
+			// We need to enforce T is a class, so if it changes that change is available in both collections.
+			valueByPlayerIndex = new T[players.Length];
+			valueByPlayer = new Dictionary<Player, T>(players.Length);
+			for (var i = 0; i < players.Length; i++)
```

Comment:
If we're going for performance in places like this shouldn't you pull out the upper limit?
`int i = 0, max = players.Length; ...`